![knockout_logo](https://github.com/aniketxdey/hms_research/assets/168318141/6f6dfca6-061b-4dbe-be88-0a1727eb909e)
_All research and materials produced during tenure as Computational Genomics Researcher at Dietlein Lab, Harvard Medical School. All code and data is property of the Computational Health Informatics Program at Harvard Medical School._

# Computational Solutions in the Tumor Microenvironment

In context of cellular cancer proliferation, the transition from healthy cells to maligant tumor cells depends on the presence few “driver” mutations in the cancer genome. These drivers serve as targets of sorts within the Tumor Microenvironment (TME), and can be exploited to produce targeted cellular therapies by combining innovations in statistics with genome biology. My work at the Dietlein Lab thus broadly revolved around developing a series of bioinformatics and machine learning tools to interpret and visualize single cell RNA sequencing data to 

 , as well as clinical tools that could be readily deployed in CHIP clinics. The study proposes the following:


Furthermore, our lab is committed to sharing results as user-friendly resources and software that can be run in the hands of non-computational scientists. We collaborate with scientists from other disciplines to apply bioinformatics methods to clinically focused questions.

Engineered a novel ML pipeline employing a random-forest algorithm to optimize predictive completion of ligand-receptor interactions with up to 89.32% accuracy in therapeutic development simulations. 

Then, I worked on clinical implementation of this method in 6+ drug discovery clinics at Boston Medical & UMass Memorial, & published my work in UMiami’s NSF-sponsored HRJ. 
