![knockout_logo](https://github.com/aniketxdey/hms_research/assets/168318141/6f6dfca6-061b-4dbe-be88-0a1727eb909e)
_All research and materials produced during tenure as Computational Genomics Researcher at Dietlein Lab, Harvard Medical School. All code and data is property of the Computational Health Informatics Program at Harvard Medical School._

# Computational Solutions in the Tumor Microenvironment
## Introduction
In the context of cellular cancer proliferation, the transition from healthy cells to maligant tumor cells depends on the presence few “driver” mutations in the cancer genome. These drivers therefore serve as targets of sorts within the Tumor Microenvironment (TME), and can be exploited to produce targeted cellular therapies by combining innovations in statistics with genome biology. Specifically, a broad range of machine learning algorithms can be utilized to classify transitions between healthy and malignant phases, as well as provide predictions for future genomic events. My research at the Dietlein Lab thus revolved around developing a series of bioinformatics tools to interpret and visualize single cell RNA sequencing data in an effort to predict genome drivers as well as targeted compounds for treatment. The study implements the following:

1) Identifying gene-expressions within the TME that drive cellular transitions towards malignant development through a variety of single-cell-omics visualizations like Leiden Clustering and differential analysis.

2) Developing a novel ML pipeline employing a random-forest algorithm to optimize predictive completion of ligand-receptor interactions at gene-expressions identified by algrotihms in (1) with up to 89.32% accuracy in therapeutic development simulations.

3) Collaborating with clinicians to share findings & models as user-friendly tutorial resources and software that can be run in the hands of non-computational scientists.
  
