![knockout_logo](https://github.com/aniketxdey/hms_research/assets/168318141/6f6dfca6-061b-4dbe-be88-0a1727eb909e)
_All research and materials produced during tenure as Computational Genomics Researcher at Dietlein Lab, Harvard Medical School. All code and data is property of the Computational Health Informatics Program at Harvard Medical School._

# Computational Solutions in the Tumor Microenvironment
## Introduction
In the complex process of cellular cancer proliferation, the transition from healthy, normal cells to malignant tumor cells is a multistep process that is driven by the accumulation of specific genetic alterations known as "driver" mutations within the cancer genome. These driver mutations are critical for the initiation and progression of cancer, as they confer selective advantages to the cells, enabling them to proliferate uncontrollably, evade apoptosis, and acquire invasive and metastatic properties. The presence of these driver mutations within the Tumor Microenvironment (TME) creates a unique genetic landscape that can be exploited for the development of targeted cellular therapies.

To effectively identify and target these driver mutations, innovative approaches that combine advancements in statistics and genome biology are being employed. Machine learning algorithms, a subset of artificial intelligence, have emerged as powerful tools for analyzing the vast amounts of genomic data generated by high-throughput sequencing technologies. These algorithms can be trained to recognize patterns and features within the cancer genome that are associated with the transition from healthy to malignant states.

My research at the Dietlein Lab thus revolved around developing a series of bioinformatics tools to interpret and visualize single cell RNA sequencing data in an effort to predict genome drivers as well as targeted compounds for treatment. The study implements the following:

1) Identifying gene-expressions within the TME that drive cellular transitions towards malignant development through a variety of unsupervised learning algorithms for single-cell-omics like Leiden Clustering and differential analysis.

2) Developing a novel ML pipeline employing a random-forest algorithm to optimize predictive completion of ligand-receptor interactions at gene-expressions identified by algrotihms in (1) with up to 89.32% accuracy in therapeutic development simulations.

3) Collaborating with clinicians to share findings & models as user-friendly tutorial resources and software that can be run in the hands of non-computational scientists.
  
## Directory Contents

- models
  - 
